lab4/viewpoints/development/development.tex

```diff
@@ -0,0 +1,131 @@
+\end{itemize}
+
+\subsubsection{Question Bank Module}
+
+\begin{itemize}
+  \item \textbf{QuestionController}: Manages interactions related to question management.
+  \item \textbf{QuestionService}: Contains the business logic for question management.
+  \item \textbf{QuestionRepository}: Responsible for storing and retrieving question data.
+\end{itemize}
+
+\subsubsection{Common Processing}
+
+\item \textbf{Data Validation:} Uses React with Formik and Yup for form handling and schema-based validation.
+
+\item \textbf{Error Handling:} Uses ASP.NET Core middleware for global exception handling, logging errors to a centralized logging service (like Serilog) and returning standardized error responses.
+
+\item \textbf{User Authentication:} Manages user authentication and authorization using JWT tokens for securing API endpoints. Middleware checks token validity and user roles before processing quiz-related requests.
+
+\begin{figure}[htbp]
+    \centering
+    \includedrawio[width=\textwidth]{viewpoints/development/development.drawio}
+    \caption{Module Organization Diagram}
+    \label{fig:development}
+\end{figure}
+
+\subsection{Standardization of Design}
+
+\subsubsection{Design Patterns}
+
+\begin{itemize}
+  \item \textbf{Factory Pattern:} Used for creating instances of different quiz types.
+  \item \textbf{Repository Pattern:} Abstracts data access, promoting a clean separation between the business logic and data access layers.
+  \item \textbf{Singleton Pattern:} Ensures a single instance of services like QuizSessionService to manage state across the application.
+\end{itemize}
+
+\subsubsection{Standard UI Components}
+
+  \item \textbf{Component Library:} Uses a custom component library based on Material-UI. Standard components like Button, Card, and Modal are reused across all quiz-related pages to maintain a consistent look and feel.
+
+\subsubsection{Consistent User Flows}
+
+\item \textbf{Quiz Taking Flow:} Users navigate from the quiz list to the quiz page, answer questions sequentially, and then view results. Each step is guided by consistent visual cues and navigation buttons.
+
+\item \textbf{Quiz Management Flow:} Admins follow a flow from creating a new quiz, adding questions, setting correct answers, and publishing the quiz.
+
+\subsection{Standardization of Testing}
+
+\subsubsection{Unit Testing}
+
+\item \textbf{Frontend:} Uses Jest and React Testing Library to test individual React components. Tests cover rendering, user interactions, and state changes.
+
+\item \textbf{Backend:} Utilizes NUnit for testing ASP.NET Core services and repositories. Tests include validation logic, business rules, and data access methods.
+
+\subsubsection{Integration Testing}
+
+\item \textbf{API Endpoints:} Uses Postman to automate testing of API endpoints.
+
+\item \textbf{Frontend-Backend Integration:} End-to-end tests with Cypress to verify that the frontend and backend interact correctly.
+
```

Comment:
We need to have a deeper testability exploration. Also provide some integration/end to end scenarios (That's the testability section from teams plan). Basically, entire test flow (test runner opens page, clicks, checks that we get response and so on). We will be implementing quiz editor, so described test must test it (aka topic owner will see quiz edit button and non owner won't, or something else).